Add optional merging and trimming

maxibor opened the pull request:
In response to #137

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [x] If necessary, also make a PR on the [nf-core/eager branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/eager)
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Make sure your code lints (`nf-core lint .`).
 - [x] Documentation in `docs` is updated
 - [ ] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated

**Learn more about contributing:** https://github.com/nf-core/eager/tree/master/.github/CONTRIBUTING.md

apeltzer commented on docs/usage.md:
````diff
@@ -162,6 +162,14 @@ A normal glob pattern, enclosed in quotation marks, can then be used for `--read
 ```
 
+### `--noCollapse`
+If you have paired-end data, but you don't want to merge them, add the commind line argument `--noCollapse`. 
````
```suggestion
If you have paired-end data, but you don't want to merge them, add the command line argument `--noCollapse`. 
```

apeltzer commented on docs/usage.md:
````diff
@@ -162,6 +162,14 @@ A normal glob pattern, enclosed in quotation marks, can then be used for `--read
+
+For example
+```bash
+--pairedEnd --noCollapse '*.fastq'
````
```suggestion
--pairedEnd --noCollapse  --reads '*.fastq'
```

apeltzer commented on main.nf:
```diff
@@ -593,19 +601,25 @@ process adapter_removal {
+    if( !params.singleEnd && !params.noCollapse){
     """
-    AdapterRemoval --file1 ${reads[0]} --file2 ${reads[1]} --basename ${prefix} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --adapter2 ${params.clip_reverse_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} --minadapteroverlap ${params.min_adap_overlap} --collapse
+    AdapterRemoval --file1 ${reads[0]} --file2 ${reads[1]} --basename ${name} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --adapter2 ${params.clip_reverse_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} --minadapteroverlap ${params.min_adap_overlap} --collapse
```
```suggestion
    AdapterRemoval --file1 ${reads[0]} --file2 ${reads[1]} --basename ${base} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --adapter2 ${params.clip_reverse_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} --minadapteroverlap ${params.min_adap_overlap} --collapse
```

apeltzer commented on main.nf:
```diff
@@ -593,19 +601,25 @@ process adapter_removal {
-    } else {
+    } else if (!params.singleEnd && params.noCollapse) {
+    """
+    AdapterRemoval --file1 ${reads[0]} --file2 ${reads[1]} --basename ${name} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --adapter2 ${params.clip_reverse_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} --minadapteroverlap ${params.min_adap_overlap}
```
```suggestion
    AdapterRemoval --file1 ${reads[0]} --file2 ${reads[1]} --basename ${base} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --adapter2 ${params.clip_reverse_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} --minadapteroverlap ${params.min_adap_overlap}
```

apeltzer commented on main.nf:
```diff
@@ -593,19 +601,25 @@ process adapter_removal {
+    AdapterRemoval --file1 ${reads[0]} --basename ${name} --gzip --threads ${task.cpus} --trimns --trimqualities --adapter1 ${params.clip_forward_adaptor} --minlength ${params.clip_readlength} --minquality ${params.clip_min_read_quality} 
     # Pseudo-Combine
-    mv *.truncated.gz ${prefix}.combined.fq.gz
+    mv *.truncated.gz ${name}.combined.fq.gz
```
```suggestion
    mv *.truncated.gz ${base}.combined.fq.gz
```

maxibor commented:
Now inivudualized to:
- skip_collapse
- skip_trim

It it now possible to:
- trim but not collapse
- collapse but not trim
- avoid collapssing and trimming alltogether (skips ar2)

I went with the "lazy" option of skipping ar2 by changing the script of the ar2 process to just a simple `mv`

jfy133 commented on main.nf:
```diff
@@ -44,6 +44,8 @@ def helpMessage() {
       --saveReference               Saves reference genome indices for later reusage
 
     Skipping                        Skip any of the mentioned steps
+      --skip_collapse               Skip merging Forward and Reverse reads together. (Only for pairedEnd samples)
```
I think it would be clearer/more consistent to have a `--skip_adapterremoval` flag here (which switches both `--skip_collapse` and `--skip_trim` to false).

Then we move the `--skip_collapse` and `--skip_trim` in help message to 'AdapterRemoval' section, as   in that section you're actually modifying the _module_, not turning entire modules on and off.

jfy133 commented on main.nf:
```diff
@@ -348,6 +356,8 @@ summary['Reads']        = params.reads
 summary['Fasta Ref']    = params.fasta
 if(params.bwa_index) summary['BWA Index'] = params.bwa_index
 summary['Data Type']    = params.singleEnd ? 'Single-End' : 'Paired-End'
+summary['Skip Collapsing'] = params.skip_collapse ? 'Yes' : 'No'
```
Is there a reason why these aren't logical/boolean? It would be more consistent with other `param.skip_*`s

maxibor replied:
These are still booleans, the yes/no is just for logging purposes.

jfy133 replied:
:+1: my misunderstanding.

apeltzer commented:
Hi @maxibor @jfy133 ! I'll go through the code asap and will adjust the testcase or figure out why it fails now. Probably something smaller, but anyways a good point to check it first 👍

apeltzer commented:
Sorry for not looking at this earlier - I'll take some time very soon to get this in and comment on the various points to get this going 👍

apeltzer commented:
I fixed the merge conflicts and we'll see if things work fine now - once this passes I think I'm already pretty happy about it - maybe some option renaming as @jfy133 suggested and some docs / changelog updates and we're almost good to go 👍

apeltzer commented:
Ok, bam input is broken now - everything else works fine - @maxibor do you want to take a look at this and maybe also have the command line switches as suggested by @jfy133 implemented?

maxibor commented:
> Ok, bam input is broken now - everything else works fine - @maxibor do you want to take a look at this and maybe also have the command line switches as suggested by @jfy133 implemented?

- Matched @jfy133 skip_* pattern suggestion
- Updated test
- Updated doc
- `Local` executor selection when `skip_adapterremoval` is set to `trye` to avoid waiting for job scheduler (SLURM, SGE, ...) to allocate ressources for a simple `mv`

jfy133 commented:
Looks good me now. Have you tested it locally yourself? I've never had any cases where I needed to skip collapse or merge so I cant see what happens downstream.

I'll try and test the skip module itself this evening.

apeltzer commented:
I guess we could actually use a `mix` operator here to get around the local executor. Might have a look too :-)

jfy133 commented:
Note to self: add to docs saying that if `--skip_adapterremoval` selected by user, the input FASTQs must be with `--singleEnd`

apeltzer commented:
Is that true? I mean it wouldn't be a big issue to make both `bwa`, `bwa aln`capable of running in PE mode then for these cases as we have the command line switch already there and present. I could even add specific tests to test this behaviour as well... ?

jfy133 commented:
> Is that true? I mean it wouldn't be a big issue to make both `bwa`, `bwa aln`capable of running in PE mode then for these cases as we have the command line switch already there and present. I could even add specific tests to test this behaviour as well... ?

Partly true. That note would admittedly be aimed more more novice users (I will switch out 'must' with 'generally will be'). But I personally can't think of any case where you would run paired end data _without_ merging/trimming. Although on second thoughts that's maybe what Alex Huebner was partly referring to with the whole modern data issue in #64 ...

At least in my experience the only time when I wanted to skip AR was when I had already processed it previously and wanted to map the merged/trimmed data to a different reference genome.

apeltzer commented:
Yes, but as we already have everything configurable now - I'm working on adding the purely modern handling (just apply trimming to reads, keep them as pairs and map them using bwa/bwa mem in PE mode) as an option in the pipeline too. This will just require speciying `--skip_collapse` when having PE data, then just trimming will be applied and mapping will happen in PE mode as one would do for modern data 👍

apeltzer commented:
https://github.com/maxibor/eager/pull/1 

Here you go - once this is merged, we can basically just add something on top of it :-)

apeltzer commented:
I merged this into the `ar_possible` branch and will merge my changes on top of this - this should make it possible for both of us to have a final check before merging to `dev` in the end!
